### PR TITLE
MPI fixes

### PR DESCRIPTION
**Pull request by jchodera**
Fixes issue noted in https://github.com/choderalab/yank-examples/pull/31#issuecomment-259518476 where error messages were not being printed before MPI Abort on error.

### CODE REVIEW

File: Yank/utils.py [hunk 1]
```diff
@@ -178,8 +178,13 @@ def initialize_mpi():
     # Override sys.excepthook to abort MPI on exception
     def mpi_excepthook(type, value, traceback):
         sys.__excepthook__(type, value, traceback)
+        sys.stdout.flush()
+        sys.stderr.flush()
         if mpicomm.size > 1:
             mpicomm.Abort(1)
+    # Store old exception handler
+    sys.__excepthook__ = sys.excepthook
```
Comment (andrrizzi) [inline Yank/utils.py]: This line is not necessary. `__excepthook__` always stores the original: https://docs.python.org/3.5/library/sys.html#sys.__excepthook__

**Comment by jchodera** [comment]
I've also added a band-aid fix for the issue noted [here](https://github.com/choderalab/yank-examples/pull/31#issuecomment-259535913) where `setup/molecules/*/` files generated last by rank 0 are not found by the other ranks if we immediately proceed through the MPI barrier after setup. This fix just adds a sleep of 5 seconds, but we might want to see if we can flush the disk buffers or improve the control logic in a longer-term fix.

**Comment by jchodera** [comment]
This still isn't fixing it.

**Comment by jchodera** [comment]
The bug is due to [this](https://github.com/choderalab/yank-examples/pull/31#issuecomment-259691959).

At some point in the future, we probably want to revamp the MPI logic to simplify things. This is low priority, but it will save us debugging time in the future, and make it easier to make our setup pipeline more extensible.

**Comment by jchodera** [comment]
@andrrizzi : I'm not actually sure how to fix this, since I realize I don't know how the expansion of your combinatorial experiments works.

I was originally thinking adding a simple `self._mpicomm.bcast()` to broadcast a data structure from rank 0 to all nodes is sufficient here, but I'm not sure the case. If you iterate through a `dict`, for example, that would potentially still produce random results---you'd need an `OrderedDict` instead.

Can you fix in this PR, @andrrizzi? I've eliminated my other erroneous changes.

**Comment by andrrizzi** [comment]
Sorry, just saw this. Of course!

**Comment by jchodera** [comment]
Thanks, @andrrizzi! Feel free to merge when you are happy with things.

**Comment by andrrizzi** [comment]
Phew! This was tricky to narrow down! During the py2 to py3 conversion the `CombinatorialLeaf` equal comparison was made orderless to handle the unordered dictionary mess, so the test that should have caught this was effectively deactivated. It should be solved now and the test is back! I'll merge as soon as tests pass.

**Comment by jchodera** [comment]
Looks like there is an issue with `test_yaml.py` where the expected and expanded combinatorial output differ, presumably because things are reordered now. I'll see if I can fix this, and also have the assertion produce a more useful error message.

**Comment by jchodera** [comment]
I'm now writing out more information about the assertion failure, re-rendering the `dict`s as YAML. Here's the failure---I don't think I'm actually able to debug this easily, so if you can look at it Friday, @andrrizzi, I'd be grateful!
```
======================================================================
FAIL: Check that combinatorial molecules are handled correctly.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/choderaj/miniconda/lib/python2.7/site-packages/nose/case.py", line 197, in runTest
    self.test(*self.arg)
  File "/Users/choderaj/github/choderalab/yank/yank.choderalab/Yank/tests/test_yaml.py", line 1003, in test_expand_molecules
    assert expanded == expected, "Expected:\n%s\nExpanded:\n%s\n" % (yaml.dump(expected), yaml.dump(expanded))
AssertionError: Expected:
experiments: {protocol: absolute-binding, system: sys}
molecules:
  lig_0_iupac1:
    epik: 0
    leap: {parameters: leaprc.gaff}
    name: iupac1
  lig_0_iupac2:
    epik: 0
    leap: {parameters: leaprc.gaff}
    name: iupac2
  lig_2_iupac1:
    epik: 2
    leap: {parameters: leaprc.gaff}
    name: iupac1
  lig_2_iupac2:
    epik: 2
    leap: {parameters: leaprc.gaff}
    name: iupac2
  mol2_0:
    filepath: /var/folders/m1/wl3ptzqx2csf2r_q2bfy78n0h7kcfv/T/tmpd7Egpt/multimol2.mol2
    leap: {parameters: leaprc.gaff}
    select: 0
  mol2_1:
    filepath: /var/folders/m1/wl3ptzqx2csf2r_q2bfy78n0h7kcfv/T/tmpd7Egpt/multimol2.mol2
    leap: {parameters: leaprc.gaff}
    select: 1
  multi_0:
    filepath: /var/folders/m1/wl3ptzqx2csf2r_q2bfy78n0h7kcfv/T/tmpd7Egpt/multipdb.pdb
    leap: {parameters: oldff/leaprc.ff14SB}
    select: 0
  multi_1:
    filepath: /var/folders/m1/wl3ptzqx2csf2r_q2bfy78n0h7kcfv/T/tmpd7Egpt/multipdb.pdb
    leap: {parameters: oldff/leaprc.ff14SB}
    select: 1
  rec_multimol2:
    filepath: /var/folders/m1/wl3ptzqx2csf2r_q2bfy78n0h7kcfv/T/tmpd7Egpt/multimol2.mol2
    leap: {parameters: oldff/leaprc.ff14SB}
  rec_multisdf:
    filepath: /var/folders/m1/wl3ptzqx2csf2r_q2bfy78n0h7kcfv/T/tmpd7Egpt/multisdf.sdf
    leap: {parameters: oldff/leaprc.ff14SB}
  sdf_0:
    filepath: /var/folders/m1/wl3ptzqx2csf2r_q2bfy78n0h7kcfv/T/tmpd7Egpt/multisdf.sdf
    leap: {parameters: leaprc.gaff}
    select: 0
  sdf_1:
    filepath: /var/folders/m1/wl3ptzqx2csf2r_q2bfy78n0h7kcfv/T/tmpd7Egpt/multisdf.sdf
    leap: {parameters: leaprc.gaff}
    select: 1
  smiles_0:
    filepath: /var/folders/m1/wl3ptzqx2csf2r_q2bfy78n0h7kcfv/T/tmpd7Egpt/multismiles.smiles
    leap: {parameters: leaprc.gaff}
    select: 0
  smiles_1:
    filepath: /var/folders/m1/wl3ptzqx2csf2r_q2bfy78n0h7kcfv/T/tmpd7Egpt/multismiles.smiles
    leap: {parameters: leaprc.gaff}
    select: 1
protocols:
  absolute-binding:
    complex:
      alchemical_path:
        lambda_electrostatics: [1.0, 0.9, 0.8, 0.6, 0.4, 0.2, 0.0]
        lambda_sterics: [1.0, 0.9, 0.8, 0.6, 0.4, 0.2, 0.0]
    solvent:
      alchemical_path:
        lambda_electrostatics: [1.0, 0.8, 0.6, 0.3, 0.0]
        lambda_sterics: [1.0, 0.8, 0.6, 0.3, 0.0]
solvents:
  solv1: {nonbonded_method: NoCutoff}
  solv2: {clearance: 10*angstroms, nonbonded_cutoff: 1*nanometer, nonbonded_method: PME}
systems:
  sys:
    ligand: !!python/object/new:yank.utils.CombinatorialLeaf
      listitems: [lig_0_iupac2, lig_2_iupac1, lig_2_iupac2, lig_0_iupac1]
    receptor: !!python/object/new:yank.utils.CombinatorialLeaf
      listitems: [rec_multisdf, rec_multimol2, multi_1, multi_0]
    solvent: !!python/object/new:yank.utils.CombinatorialLeaf
      listitems: [solv1, solv2]

Expanded:
!!python/object:yank.utils.CombinatorialTree
_d:
  experiments: {protocol: absolute-binding, system: sys}
  molecules:
    lig_0_iupac1:
      epik: 0
      leap: {parameters: leaprc.gaff}
      name: iupac1
    lig_0_iupac2:
      epik: 0
      leap: {parameters: leaprc.gaff}
      name: iupac2
    lig_2_iupac1:
      epik: 2
      leap: {parameters: leaprc.gaff}
      name: iupac1
    lig_2_iupac2:
      epik: 2
      leap: {parameters: leaprc.gaff}
      name: iupac2
    mol2_0:
      filepath: /var/folders/m1/wl3ptzqx2csf2r_q2bfy78n0h7kcfv/T/tmpd7Egpt/multimol2.mol2
      leap: {parameters: leaprc.gaff}
      select: 0
    mol2_1:
      filepath: /var/folders/m1/wl3ptzqx2csf2r_q2bfy78n0h7kcfv/T/tmpd7Egpt/multimol2.mol2
      leap: {parameters: leaprc.gaff}
      select: 1
    multi_0:
      filepath: /var/folders/m1/wl3ptzqx2csf2r_q2bfy78n0h7kcfv/T/tmpd7Egpt/multipdb.pdb
      leap: {parameters: oldff/leaprc.ff14SB}
      select: 0
    multi_1:
      filepath: /var/folders/m1/wl3ptzqx2csf2r_q2bfy78n0h7kcfv/T/tmpd7Egpt/multipdb.pdb
      leap: {parameters: oldff/leaprc.ff14SB}
      select: 1
    rec_multimol2:
      filepath: /var/folders/m1/wl3ptzqx2csf2r_q2bfy78n0h7kcfv/T/tmpd7Egpt/multimol2.mol2
      leap: {parameters: oldff/leaprc.ff14SB}
    rec_multisdf:
      filepath: /var/folders/m1/wl3ptzqx2csf2r_q2bfy78n0h7kcfv/T/tmpd7Egpt/multisdf.sdf
      leap: {parameters: oldff/leaprc.ff14SB}
    sdf_0:
      filepath: /var/folders/m1/wl3ptzqx2csf2r_q2bfy78n0h7kcfv/T/tmpd7Egpt/multisdf.sdf
      leap: {parameters: leaprc.gaff}
      select: 0
    sdf_1:
      filepath: /var/folders/m1/wl3ptzqx2csf2r_q2bfy78n0h7kcfv/T/tmpd7Egpt/multisdf.sdf
      leap: {parameters: leaprc.gaff}
      select: 1
    smiles_0:
      filepath: /var/folders/m1/wl3ptzqx2csf2r_q2bfy78n0h7kcfv/T/tmpd7Egpt/multismiles.smiles
      leap: {parameters: leaprc.gaff}
      select: 0
    smiles_1:
      filepath: /var/folders/m1/wl3ptzqx2csf2r_q2bfy78n0h7kcfv/T/tmpd7Egpt/multismiles.smiles
      leap: {parameters: leaprc.gaff}
      select: 1
  protocols:
    absolute-binding:
      complex:
        alchemical_path:
          lambda_electrostatics: [1.0, 0.9, 0.8, 0.6, 0.4, 0.2, 0.0]
          lambda_sterics: [1.0, 0.9, 0.8, 0.6, 0.4, 0.2, 0.0]
      solvent:
        alchemical_path:
          lambda_electrostatics: [1.0, 0.8, 0.6, 0.3, 0.0]
          lambda_sterics: [1.0, 0.8, 0.6, 0.3, 0.0]
  solvents:
    solv1: {nonbonded_method: NoCutoff}
    solv2: {clearance: 10*angstroms, nonbonded_cutoff: 1*nanometer, nonbonded_method: PME}
  systems:
    sys:
      ligand: !!python/object/new:yank.utils.CombinatorialLeaf
        listitems: [lig_0_iupac1, lig_0_iupac2, lig_2_iupac1, lig_2_iupac2]
      receptor: !!python/object/new:yank.utils.CombinatorialLeaf
        listitems: [rec_multimol2, rec_multisdf, multi_0, multi_1]
      solvent: !!python/object/new:yank.utils.CombinatorialLeaf
        listitems: [solv1, solv2]



----------------------------------------------------------------------
Ran 125 tests in 164.977s
```

**Comment by andrrizzi** [comment]
Fixed! Merging now.

**Pull request merged**